fix(lsp): fix some small bugs in snippet-parser

L3MON4D3 opened the pull request:
As mentioned in #18772, there are some small bugs in the current parser.

I've been able to fix the parsing of `the format`(`${VARIABLE/regex/format}}`)-variants (`:else_text`, `:-else_text`), but handling emtpy `if/else_text` (and empty placeholder text, which is also technically allowed) seems harder:
Accepting an empty string in `take_until` leads to infinite loops in [`P.many`](https://github.com/L3MON4D3/neovim/blob/cb1717d9639bd947c09cc44076bc808c4f2aba0d/runtime/lua/vim/lsp/_snippet.lua#L112-L117), and I have no idea how to prevent those.

(Empty strings are technically allowed, but any snippet that includes an empty string can be written without it (`${1:if_text?}` -> `${1:+if_text}`, `${1:}` -> `$1`))

(vscode seems to handle empty strings just fine, but IIRC they also use a parser combinator, so I'll have to look into that)

L3MON4D3 commented:
Added a field to prevent empty strings, setting it after receiving an empty string in `many` allows empty strings in e.g. placeholders(`${1:}).
It's not a good solution though, now there are unnecessary empty strings in some ASTs, since `many` is greedy and can always match an additional.
Might be enough to just remove empty strings if they are not the only child, but I'd rather implement a better mechanism that does not require this clean-up.

hrsh7th commented on test/functional/plugin/lsp/snippet_spec.lua:
```diff
@@ -148,5 +148,47 @@ describe('vim.lsp._snippet', function()
+                type = snippet.NodeType.FORMAT,
+                capture_index = 1,
+				else_text = 'else_text',
+              },
```
Sorry. Actually, I don't know the meaning of `':+' if` like format.

In my eyes, this parsed results does not contain the `+` or `-` or `?` context. Is this expected?

(Anyway, Thank you for your fix!)

hrsh7th replied:
```lua
S.format = P.any(
  P.map(P.seq(S.dollar, S.int), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[2],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, S.slash, P.any(
    P.token('upcase'),
    P.token('downcase'),
    P.token('capitalize'),
    P.token('camelcase'),
    P.token('pascalcase')
  ), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      modifier = values[6],
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.question, P.take_until({ ':' }, { '\\' }), S.colon, P.take_until({ '}' }, { '\\' })),  S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      if_text = values[5][2].esc,
      else_text = (values[5][4] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.plus, P.opt(P.take_until({ '}' }, { '\\' }))), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      if_text = (values[5][2] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.minus, P.opt(P.take_until({ '}' }, { '\\' }))), S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      else_text = (values[5][2] or {}).esc,
    }, Node)
  end),
  P.map(P.seq(S.dollar, S.open, S.int, S.colon, P.opt(P.take_until({ '}' }, { '\\' })) ,  S.close), function(values)
    return setmetatable({
      type = Node.Type.FORMAT,
      capture_index = values[3],
      else_text = (values[5] or {}).esc,
    }, Node)
  end)
)
```

In my local environment, the above parser can pass your tests without modifying the parser combinators. 👍🏼

L3MON4D3 replied:
Thank you for taking a look :)

> OK. VSCodes doesn't have context of + and -. It's just a kind of if/else value.

Yeah, just `if_text` for existing/nonemtpy capture and `else_text` otherwise, at least that's what I got when playing around with it in vscode.

>In my local environment, the above parser can pass your tests without modifying the parser combinators. 👍🏼

Right, the parser combinator-modifications are necessary for handling empty strings, I'll add a few tests for those (and get this ready for review in general :sweat_smile:)

hrsh7th replied:
```lua
parse('${VAR/regex/${1:/upcase}${1:+if_text}${1:-else_text}${1:?if_text:else_text}${1:else_text}${1:+}${1:-}${1:?:}${1:}/}')
```

My local code was passed the above test case without parser combinator-modifications :)
I think you should use `P.opt` instead of `prevent_empty`.

hrsh7th replied:
Note: If you want to treat the omitted value as `''` (empty string). You can convert `else_text = (values[5] or {}).esc or ''`

L3MON4D3 replied:
> ```lua
> parse('${VAR/regex/${1:/upcase}${1:+if_text}${1:-else_text}${1:?if_text:else_text}${1:else_text}${1:+}${1:-}${1:?:}${1:}/}')
> ```
> 
> My local code was passed the above test case without parser combinator-modifications :) I think you should use `P.opt` instead of `prevent_empty`.

Ahhh sorry, I didn't take that close a look at it.
`opt` is indeed much better👍👍

L3MON4D3 committed:
fix(lsp): fix small bugs in snippet-parser

This fixes the following bugs:
`${1:else_text}` -> format with if_text: "else_text"
`${1:-else_text}` -> format with if_text: "else_text"
`${1:}` in `format` (eg. empty else_text) -> error.
`${1:}` (eg. empty placeholder) -> error.

Thanks @hrsh7th :)

L3MON4D3 commented on runtime/lua/vim/lsp/_snippet.lua:
```diff
@@ -156,10 +156,10 @@ P.seq = function(...)
     local new_pos = pos
-    for _, parser in ipairs(parsers) do
+    for i, parser in ipairs(parsers) do
       local result = parser(input, new_pos)
       if result.parsed then
-        table.insert(values, result.value)
+        values[i] = result.value
```
Previously empty `opt` were just omitted in `seq` (`nil` as last item in list is ignored), now they're `nil`.
I don't _think_ this breaks anything, wanted to highlight it nonetheless since I'm not 100% certain.
Do you think this could cause any issues @hrsh7th?

hrsh7th replied:
I think this fix is make sense. Thank you! 👍🏼

hrsh7th commented on runtime/lua/vim/lsp/_snippet.lua:
```diff
@@ -272,22 +272,48 @@ S.format = P.any(
+    end
+  ),
+  P.map(
+    P.seq(S.dollar, S.open, S.int, S.colon, P.seq(S.minus, P.opt(P.take_until({ '}' }, { '\\' }))), S.close),
```
Ah... Sorry. It can be more simple. I think nested P.seq might not needed.
